theano/gof/opt.py

```diff
@@ -858,13 +858,40 @@ def apply(self, fgraph):
                                 hasattr(c.op, 'destroy_map')]) > 1:
                             continue
 
+                if pairs[0][0].type != pairs[0][1].type:
```

Comment:
Using convert_variable allow to simplify heavily the code as it make many of the check already. Now all this if can become:
```
if len(pairs) == 1 and pairs[0][0].type != pairs[0][1].type:
   res = ...convert_variable(...
   if res:
      pairs = ...
   else:
       res = ...convert_variable() in the other direction
       if res:
           pairs = ...
```

this will keep the original error.